### PR TITLE
Product Selector: Add filter functionality

### PR DESCRIPTION
**Pull request by itsmeichigo**
Part of #6489 

### Description

This PR continues the customization of the product selector to add a filter bar on top of the product list:
- Add a bar with two new buttons: Select All and Filter.
  - The Select All button should only be visible when multiple selections are enabled and when there are results present on the view.
  - The Filter is available when the `showsFilter` configuration is enabled for the product selector.
- Handle filter functionality:
  - Wrap `FilterListViewController` in a SwiftUI view to be presented from the product selector.
  - Update `ProductSelectorViewModel` to update the product results controller when filters are updated.
  - Update the title of the filter button when the number of active filters changes.
- Enable filter for `AddEditCoupon` and `CouponRestrictions`.

### Testing instructions

Testing coupon edit and coupon restrictions:
- Make sure that your test store has coupons enabled and has set up at least one coupon in WP Admin > Marketing > Coupons.
- Open the app, make sure that coupon management is enabled in Menu > Settings > Experimental Features > Coupon Management.
- Navigate to Menu > Coupons and select a coupon.
- Tap the ellipsis button > Edit Coupon > Select the All Products/Edit Products button.
- Notice that 2 buttons Select All and Filter are present.
- Tap the Filter button, update the filters to your liking, and apply the filter. The product list should show only the matching items, and the Filter button should display the correct number of active filters.
- If no result is found, the Filter button should still be present while the Select All button is not.

Please feel free to test the product selector on the New Order screen to make sure that single selection still works as before.

### Screenshots

https://user-images.githubusercontent.com/5533851/164704470-f7737cd8-e7fd-4c62-ac7f-8a3906e651be.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by pmusolino — COMMENTED** [review]
I tested it, and the only thing that it seems to not work properly is the category filter. For some reason, it doesn't care what category I select, the list of products will be not updated.

**Some extra things not strictly related to this PR:**

It looks like that when the title of a product goes on multiple lines, the alignment of the text is centered and not leading.



If I select a product with variations, also if I select all the variations available, the icon of the main product will be a `minus` and not a `check` (see the "Patient Ninja" product). Is it correct?

**Comment by itsmeichigo** [comment]
Thank you @pmusolino for noticing the issues!

As for the issue with filtering product categories, this was due to me not sending the filters when resynching products. I updated this in 1029b93a5137b686dd11d2f16865274f3244ecff.

There's another edge case about searching product with filters, this requires updating Network and Yosemite layers so I added that in a separate PR #6740.

> It looks like that when the title of a product goes on multiple lines, the alignment of the text is centered and not leading.

I fixed this in 1e1a3d4d183152665aa2fc990d650c1901a3f4b9.

> If I select a product with variations, also if I select all the variations available, the icon of the main product will be a minus and not a check (see the "Patient Ninja" product). Is it correct?

The reason for this is that products and variations are currently filtered to show only purchasable items. I added a quick fix in ebb8ee602bc87818487ef02cfff85e6c607ff1d7 to only apply this filter in the Order Creation flow. I believe we need a better fix just in case we need to select multiple purchasable items - this should be fixed in a separate issue.

I also added an update to clear both search queries and filters when the product selector is closed in 82013e61e0307cd60adbf3578063567e6ff37ea7 and 936795fe403f94a5d18efd5f9538e965644b23da - to make sure that the product list (on the Products tab) shows the full list of products again. This is necessary since we are using the local storage as the same single source of truth for both screens. The `onDisappear` modifier needs to be called outside of `ProductSelector` to avoid reloading the view when opening the product variation selector.

This PR is ready for another look 🙇‍♀️

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelector.swift [hunk 1]
```diff
@@ -18,6 +18,8 @@ struct ProductSelector: View {
     ///
     @Environment(\.safeAreaInsets) private var safeAreaInsets: EdgeInsets
 
+    @State private var showingFilter: Bool = false
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelector.swift]:
```suggestion
    @State private var showingFilters: Bool = false
```

File: WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelector.swift [hunk 2]
```diff
@@ -127,6 +156,7 @@ struct ProductSelector: View {
 
 extension ProductSelector {
     struct Configuration {
+        var showsFilter: Bool = false
```
Comment (pmusolino) [inline WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelector.swift]:
```suggestion
        var showsFilters: Bool = false
```

**Review by pmusolino — APPROVED** [review]
Looks good to me :shipit: thanks for fixing the things!

**Pull request merged**